### PR TITLE
Add Prometheus metrics and GoKit logs

### PR DESCRIPTION
**Pull request by tomwilkie**

### CODE REVIEW

File: events.go [hunk 1]
```diff
@@ -189,16 +189,14 @@ func (s *Session) addNewNode(host *HostInfo) {
 }
 
 func (s *Session) handleNewNode(ip net.IP, port int, waitForBinary bool) {
-	if gocqlDebug {
-		Logger.Printf("gocql: Session.handleNewNode: %s:%d\n", ip.String(), port)
-	}
+	level.Error(s.logger).Log("msg", "Session.handleNewNode", "ip", ip.String(), "port", port)
```
Comment (pracucci) [inline events.go]: Why is this error level? Looks more an Info to me.

File: events.go [hunk 2]
```diff
@@ -248,9 +244,7 @@ func (s *Session) handleRemovedNode(ip net.IP, port int) {
 }
 
 func (s *Session) handleNodeUp(eventIp net.IP, eventPort int, waitForBinary bool) {
-	if gocqlDebug {
-		Logger.Printf("gocql: Session.handleNodeUp: %s:%d\n", eventIp.String(), eventPort)
-	}
+	level.Error(s.logger).Log("msg", "Session.handleNodeUp", "ip", eventIp.String(), "port", eventPort)
```
Comment (pracucci) [inline events.go]: Why is this an Error level? Looks more an Info to me.

File: events.go [hunk 3]
```diff
@@ -274,9 +268,7 @@ func (s *Session) handleNodeUp(eventIp net.IP, eventPort int, waitForBinary bool
 }
 
 func (s *Session) handleNodeDown(ip net.IP, port int) {
-	if gocqlDebug {
-		Logger.Printf("gocql: Session.handleNodeDown: %s:%d\n", ip.String(), port)
-	}
+	level.Error(s.logger).Log("msg", "Session.handleNodeDown", "ip", ip.String(), "port", port)
```
Comment (pracucci) [inline events.go]: Error or Warn here?

File: policies.go [hunk 4]
```diff
@@ -246,10 +249,8 @@ func (d *DowngradingConsistencyRetryPolicy) Attempt(q RetryableQuery) bool {
 		return false
 	} else if currentAttempt > 0 {
 		q.SetConsistency(d.ConsistencyLevelsToTry[currentAttempt-1])
-		if gocqlDebug {
-			Logger.Printf("%T: set consistency to %q\n",
-				d,
-				d.ConsistencyLevelsToTry[currentAttempt-1])
+		if d.logger != nil {
+			level.Error(d.logger).Log("msg", "DowngradingConsistencyRetryPolicy set consistency", "consistency", d.ConsistencyLevelsToTry[currentAttempt-1])
```
Comment (pracucci) [inline policies.go]: Why Error level?

File: policies.go [hunk 5]
```diff
@@ -464,7 +465,7 @@ func (t *tokenAwareHostPolicy) SetPartitioner(partitioner string) {
 	t.mu.Lock()
 	defer t.mu.Unlock()
 
-	if t.partitioner != partitioner {
+	if t.partitioner == partitioner {
```
Comment (pracucci) [inline policies.go]: I honestly don't understand this change. Looks it introduces a bug to me.

File: policies.go [hunk 6]
```diff
@@ -550,21 +551,21 @@ func (t *tokenAwareHostPolicy) getMetadataForUpdate() *clusterMeta {
 
 // resetTokenRing creates a new tokenRing.
 // It must be called with t.mu locked.
-func (m *clusterMeta) resetTokenRing(partitioner string, hosts []*HostInfo) {
+func (m *clusterMeta) resetTokenRing(partitioner string, hosts []*HostInfo) error {
```
Comment (pracucci) [inline policies.go:554]: Now returns an error, but it's never checked in any place where we call `resetTokenRing()`

**Review by pracucci — COMMENTED** [review]
LGTM. I left few nits plus a couple of questions about changes I'm bit dubious about.

File: policies.go (hunk 5 above)
Comment (tomwilkie) [reply policies.go]: Yeah this is a mistake, good catch.

File: policies.go (hunk 6 above)
Comment (tomwilkie) [reply policies.go:554]: Agree, its tech debt I need to tidy up in a future refactor.  Shouldn't affect us.

**Pull request merged**